pytket/extensions/cutensornet/structured_state/classical.py

```diff
@@ -74,6 +98,81 @@ def apply_classical_command(
         raise NotImplementedError(f"Commands of type {op.type} are not supported.")
 
 
+def evaluate_clexpr(
+    expr: ClExpr, bitvar_val: dict[int, int], regvar_val: dict[int, int]
+) -> int:
+    """Recursive evaluation of a ClExpr."""
+
+    # Evaluate arguments to operation
+    args_val = []
+    for arg in expr.args:
+        if isinstance(arg, int):
+            value = arg
+        elif isinstance(arg, ClBitVar):
+            value = bitvar_val[arg.index]
+        elif isinstance(arg, ClRegVar):
+            value = regvar_val[arg.index]
+        elif isinstance(arg, ClExpr):
+            value = evaluate_clexpr(arg, bitvar_val, regvar_val)
+        else:
+            raise Exception(f"Unrecognised argument type of ClExpr: {type(arg)}.")
+
+        args_val.append(value)
+
+    # Apply the operation at the root of this ClExpr
+    if expr.op in [ClOp.BitAnd, ClOp.RegAnd]:
+        result = args_val[0] & args_val[1]
+    elif expr.op in [ClOp.BitOr, ClOp.RegOr]:
+        result = args_val[0] | args_val[1]
+    elif expr.op in [ClOp.BitXor, ClOp.RegXor]:
+        result = args_val[0] ^ args_val[1]
+    elif expr.op in [ClOp.BitEq, ClOp.RegEq]:
+        result = int(args_val[0] == args_val[1])
+    elif expr.op in [ClOp.BitNeq, ClOp.RegNeq]:
+        result = int(args_val[0] != args_val[1])
+    elif expr.op == ClOp.RegGeq:
+        result = int(args_val[0] >= args_val[1])
+    elif expr.op == ClOp.RegGt:
+        result = int(args_val[0] > args_val[1])
+    elif expr.op == ClOp.RegLeq:
+        result = int(args_val[0] <= args_val[1])
+    elif expr.op == ClOp.RegLt:
+        result = int(args_val[0] < args_val[1])
+    elif expr.op == ClOp.BitNot:
+        result = 1 - args_val[0]
+    # elif expr.op == ClOp.RegNot:
+    #     result = int(args_val[0] == 0)
+    elif expr.op in [ClOp.BitZero, ClOp.RegZero]:
+        result = 0
+    elif expr.op in [ClOp.BitOne, ClOp.RegOne]:
+        result = 1
+    # elif expr.op == ClOp.RegAdd:
+    #     result = args_val[0] + args_val[1]
+    # elif expr.op == ClOp.RegSub:
+    #     result = args_val[0] - args_val[1]
+    # elif expr.op == ClOp.RegMul:
+    #     result = args_val[0] * args_val[1]
+    # elif expr.op == ClOp.RegPow:
+    #     result = int(args_val[0] ** args_val[1])
+    elif expr.op == ClOp.RegRsh:
+        result = args_val[0] >> args_val[1]
+    # elif expr.op == ClOp.RegNeg:
+    #     result = -args_val[0]
+    else:
+        # TODO: Currently not supporting ClOp's RegDiv since it does not return int,
+        # so I am unsure what the semantic is meant to be.
+        # TODO: I don't now what to do with RegNot, since input
```

Comment:
The intended semantics of `RegNot` is bitwise-NOT, i.e. every bit of the register is flipped.